EasyPost/Calculation/Rates.cs

```diff
@@ -31,37 +31,87 @@ public static Rate GetLowestObjectRate(IEnumerable<Rate> rates, List<string>? in
 
             Rate? lowestRate = null;
 
-            foreach (Rate rate in from rate in rates where includeCarriers.Count <= 0 || includeCarriers.Contains(rate.Carrier.ToLower()) where excludeCarriers.Count <= 0 || !excludeCarriers.Contains(rate.Carrier.ToLower()) where includeServices.Count <= 0 || includeServices.Contains(rate.Service.ToLower()) where excludeServices.Count <= 0 || !excludeServices.Contains(rate.Service.ToLower()) select rate)
+            foreach (Rate rate in rates)
             {
-                // if lowest rate is null, set it to this rate
+                if (includeCarriers.Count > 0 || excludeCarriers.Count > 0)
+                {
+                    // we have a carrier filter
+
+                    if (rate.Carrier == null)
+                    {
+                        // If we are filtering by carrier and the rate doesn't have a carrier, skip it
+                        continue;
+                    }
+
+                    if (includeCarriers.Count > 0 && !includeCarriers.Contains(rate.Carrier!.ToLower()))
+                    {
+                        // If we have a list of carriers to include and the rate's carrier isn't in the list, skip it
+                        continue;
+                    }
+
+                    if (excludeCarriers.Contains(rate.Carrier!.ToLower()))
+                    {
+                        // If the rate's carrier is in the list of carriers to exclude, skip it
+                        continue;
+                    }
+                }
+
+                if (includeServices.Count > 0 || excludeServices.Count > 0)
+                {
+                    // we have a service filter
+
+                    if (rate.Service == null)
+                    {
+                        // If we are filtering by service and the rate doesn't have a service, skip it
+                        continue;
+                    }
+
+                    if (includeServices.Count > 0 && !includeServices.Contains(rate.Service!.ToLower()))
+                    {
+                        // If we have a list of services to include and the rate's service isn't in the list, skip it
+                        continue;
+                    }
+
+                    if (excludeServices.Contains(rate.Service!.ToLower()))
+                    {
+                        // If the rate's service is in the list of services to exclude, skip it
+                        continue;
+                    }
+                }
+
                 if (lowestRate == null)
                 {
+                    // if lowest rate is null, set it to this rate
                     lowestRate = rate;
                     continue;
                 }
 
                 if (rate.Price == null || lowestRate.Price == null)
                 {
+                    // if the rate or lowest rate doesn't have a price, throw an exception
                     throw new FilteringError(Constants.ErrorMessages.NullObjectComparison);
```

Comment:
I don't think it makes sense to use `NullObjectComparsion` for the error message when the price is null since we are not comparing objects here, `NoObjectFound` should be used instead

Reply:
It's because the price is null, and we can't compare a null price to something else, i.e. NullObjectComparison.

Changing this would introduce a breaking change if users are expecting a specific error message. We'll keep this as is for now.

Reply:
Sure, we should have caught this in the initial thread-safe rewrite